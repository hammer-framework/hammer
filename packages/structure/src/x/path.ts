import { basename, sep, normalize } from 'path'

export function directoryNameResolver(dirName: string): string {
  dirName = normalize(dirName)
  const parts = dirName.split(sep)
  const pp = parts[parts.length - 1]
  parts.push(pp)
  return parts.join(sep) + '.js'
}

export function followsDirNameConvention(filePath: string): boolean {
  filePath = normalize(filePath)
  const ending = basenameNoExt(filePath) + sep + basename(filePath)
  return filePath.endsWith(ending)
}

export function basenameNoExt(path: string): string {
  path = normalize(path)
  const parts = basename(path).split('.')
  if (parts.length > 1) parts.pop()
  return parts.join('.')
}

export function isLayoutFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Layout')
}

export function isCellFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Cell')
}
