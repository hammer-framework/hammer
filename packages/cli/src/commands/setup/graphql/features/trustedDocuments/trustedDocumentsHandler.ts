import fs from 'node:fs'
import path from 'node:path'

import toml from '@iarna/toml'
import execa from 'execa'
import { Listr } from 'listr2'
import { format } from 'prettier'

import { prettierOptions } from '@redwoodjs/cli-helpers'
import { getConfigPath, getPaths, resolveFile } from '@redwoodjs/project-config'

import { runTransform } from '../fragments/runTransform'

function updateRedwoodToml(redwoodTomlPath: string) {
  const originalTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')
  const redwoodTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')
  // Can't type toml.parse because this PR has not been included in a released yet
  // https://github.com/iarna/iarna-toml/commit/5a89e6e65281e4544e23d3dbaf9e8428ed8140e9
  const redwoodTomlObject = toml.parse(redwoodTomlContent) as any
  const hasExistingGraphqlSection = !!redwoodTomlObject?.graphql

  if (redwoodTomlObject?.graphql?.trustedDocuments) {
    console.info(
      'GraphQL Trusted Documents are already enabled in your Redwood project.'
    )

    return { newConfig: undefined, trustedDocumentsExists: true }
  }

  let newTomlContent =
    originalTomlContent.replace(/\n$/, '') +
    '\n\n[graphql]\n  trustedDocuments = true'

  if (hasExistingGraphqlSection) {
    const existingGraphqlSetting = Object.keys(redwoodTomlObject.graphql)

    let inGraphqlSection = false
    let indentation = ''
    let lastGraphqlSettingIndex = 0

    const tomlLines = originalTomlContent.split('\n')
    tomlLines.forEach((line, index) => {
      if (line.startsWith('[graphql]')) {
        inGraphqlSection = true
        lastGraphqlSettingIndex = index
      } else {
        if (/^\s*\[/.test(line)) {
          inGraphqlSection = false
        }
      }

      if (inGraphqlSection) {
        const matches = line.match(
          new RegExp(`^(\\s*)(${existingGraphqlSetting})\\s*=`, 'i')
        )

        if (matches) {
          indentation = matches[1]
        }

        if (/^\s*\w+\s*=/.test(line)) {
          lastGraphqlSettingIndex = index
        }
      }
    })

    tomlLines.splice(
      lastGraphqlSettingIndex + 1,
      0,
      `${indentation}trustedDocuments = true`
    )

    newTomlContent = tomlLines.join('\n')
  }

  return { newConfig: newTomlContent, trustedDocumentsExists: false }
}

export async function handler({ force }: { force: boolean }) {
  const tasks = new Listr(
    [
      {
        title:
          'Update Redwood Project Configuration to enable GraphQL Trusted Documents ...',
        task: () => {
          const redwoodTomlPath = getConfigPath()

          const { newConfig, trustedDocumentsExists } =
            updateRedwoodToml(redwoodTomlPath)

          if (newConfig && (force || !trustedDocumentsExists)) {
            fs.writeFileSync(redwoodTomlPath, newConfig, 'utf-8')
          }
        },
      },
      {
        title: 'Generating Trusted Documents store ...',
        task: () => {
          execa.commandSync('yarn redwood generate types', { stdio: 'ignore' })
        },
      },
      {
        title:
          'Configuring the GraphQL Handler to use a Trusted Documents store ...',
        task: async () => {
          const graphqlPath = resolveFile(
            path.join(getPaths().api.functions, 'graphql')
          )

          if (!graphqlPath) {
            throw new Error('Could not find a GraphQL handler in your project.')
          }

          const transformResult = await runTransform({
            transformPath: path.join(__dirname, 'graphqlTransform.js'),
            targetPaths: [graphqlPath],
          })

          if (transformResult.error) {
            throw new Error(transformResult.error)
          }

          const source = fs.readFileSync(graphqlPath, 'utf-8')

          const prettifiedApp = format(source, {
            ...prettierOptions(),
            parser: 'babel-ts',
          })

          fs.writeFileSync(graphqlPath, prettifiedApp, 'utf-8')
        },
      },
    ],
    { rendererOptions: { collapseSubtasks: false } }
  )

  try {
    await tasks.run()
  } catch (e: any) {
    console.error(e.message)
    process.exit(e?.exitCode || 1)
  }
}
