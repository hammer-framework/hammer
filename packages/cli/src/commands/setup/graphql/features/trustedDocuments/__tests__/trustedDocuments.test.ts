let mockExecutedTaskTitles: Array<string> = []
let mockSkippedTaskTitles: Array<string> = []

jest.mock('fs', () => require('memfs').fs)
jest.mock('node:fs', () => require('memfs').fs)
jest.mock('execa')
// The jscodeshift parts are tested by another test
jest.mock('../../fragments/runTransform', () => {
  return {
    runTransform: () => {
      return {}
    },
  }
})

jest.mock('listr2', () => {
  return {
    // Return a constructor function, since we're calling `new` on Listr
    Listr: jest.fn().mockImplementation((tasks: Array<any>) => {
      return {
        run: async () => {
          mockExecutedTaskTitles = []
          mockSkippedTaskTitles = []

          for (const task of tasks) {
            const skip =
              typeof task.skip === 'function' ? task.skip : () => task.skip

            if (skip()) {
              mockSkippedTaskTitles.push(task.title)
            } else {
              mockExecutedTaskTitles.push(task.title)
              await task.task()
            }
          }
        },
      }
    }),
  }
})

import path from 'node:path'

import { vol } from 'memfs'

import { handler } from '../trustedDocumentsHandler'

// Set up RWJS_CWD
let original_RWJS_CWD: string | undefined
const APP_PATH = '/redwood-app'

const tomlFixtures: Record<string, string> = {}

beforeAll(() => {
  original_RWJS_CWD = process.env.RWJS_CWD
  process.env.RWJS_CWD = APP_PATH

  const actualFs = jest.requireActual('fs')
  const tomlFixturesPath = path.join(__dirname, '__fixtures__', 'toml')

  tomlFixtures.default = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'default.toml'),
    'utf-8'
  )

  tomlFixtures.fragments = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'fragments.toml'),
    'utf-8'
  )

  tomlFixtures.fragmentsNoSpaceEquals = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'fragments_no_space_equals.toml'),
    'utf-8'
  )

  tomlFixtures.trustedDocsAlreadySetup = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'trusted_docs_already_setup.toml'),
    'utf-8'
  )

  tomlFixtures.trustedDocsNoSpaceEquals = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'trusted_docs_no_space_equals.toml'),
    'utf-8'
  )

  tomlFixtures.trustedDocsFragmentsAlreadySetup = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'trusted_docs_fragments_already_setup.toml'),
    'utf-8'
  )

  tomlFixtures.trustedDocsCommentedGraphql = actualFs.readFileSync(
    path.join(tomlFixturesPath, 'trusted_docs_commented_graphql.toml'),
    'utf-8'
  )
})

afterAll(() => {
  process.env.RWJS_CWD = original_RWJS_CWD
  jest.resetAllMocks()
  jest.resetModules()
})

// Silence console.info
console.info = jest.fn()

describe('Trusted documents setup', () => {
  it('runs all tasks', async () => {
    vol.fromJSON(
      { 'redwood.toml': '', 'api/src/functions/graphql.js': '' },
      APP_PATH
    )

    await handler({ force: false })

    expect(mockExecutedTaskTitles).toMatchInlineSnapshot(`
    [
      "Update Redwood Project Configuration to enable GraphQL Trusted Documents ...",
      "Generating Trusted Documents store ...",
      "Configuring the GraphQL Handler to use a Trusted Documents store ...",
    ]
  `)
  })

  describe('Project toml configuration updates', () => {
    describe('default toml where no graphql or trusted documents is setup', () => {
      it('updates the toml file with graphql and trusted documents enabled', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.default,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toMatchSnapshot()
      })
    })
    describe('default toml where graphql fragments are already setup', () => {
      it('updates the toml file with graphql and trusted documents enabled and keeps fragments', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.fragments,
            'api/src/functions/graphql.ts': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toMatchSnapshot()
      })
    })
    describe('default toml where graphql fragments are already setup using no spaces', () => {
      it('updates the toml file with graphql and trusted documents enabled and keeps fragments', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.fragmentsNoSpaceEquals,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toMatchSnapshot()
      })
    })
    describe('default toml where graphql trusted documents are already setup', () => {
      it('makes no changes as trusted documents are already setup', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.trustedDocsAlreadySetup,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toEqual(
          tomlFixtures.trustedDocsAlreadySetup
        )
      })
    })
    describe('default toml where graphql trusted documents are already setup using no spaces', () => {
      it('makes no changes as trusted documents are already setup', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.trustedDocsNoSpaceEquals,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toEqual(
          tomlFixtures.trustedDocsNoSpaceEquals
        )
      })
    })
    describe('default toml where graphql trusted documents and fragments are already setup', () => {
      it('makes no changes as trusted documents are already setup', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.trustedDocsFragmentsAlreadySetup,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toEqual(
          tomlFixtures.trustedDocsFragmentsAlreadySetup
        )
      })
    })
    describe('toml where graphql section is commented out', () => {
      it('adds a new section with `trustedDocuments = true`', async () => {
        vol.fromJSON(
          {
            'redwood.toml': tomlFixtures.trustedDocsCommentedGraphql,
            'api/src/functions/graphql.js': '',
          },
          APP_PATH
        )

        await handler({ force: false })

        expect(vol.toJSON()[APP_PATH + '/redwood.toml']).toMatchSnapshot()
      })
    })
  })
})
