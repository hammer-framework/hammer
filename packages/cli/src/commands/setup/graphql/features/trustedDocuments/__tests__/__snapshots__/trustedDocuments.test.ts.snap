// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trusted documents setup tests GraphQL Handler updates default handler where the trusted document store is not configured updates the handler with the trusted document store 1`] = `
"import { createAuthDecoder } from '@redwoodjs/auth-dbauth-api'
import { createGraphQLHandler } from '@redwoodjs/graphql-server'

import directives from 'src/directives/**/*.{js,ts}'
import sdls from 'src/graphql/**/*.sdl.{js,ts}'
import services from 'src/services/**/*.{js,ts}'

import { cookieName, getCurrentUser } from 'src/lib/auth'
import { db } from 'src/lib/db'
import { logger } from 'src/lib/logger'
import { store } from 'src/lib/trustedDocumentsStore'

const authDecoder = createAuthDecoder(cookieName)

export const handler = createGraphQLHandler({
  authDecoder,
  getCurrentUser,
  loggerConfig: { logger, options: {} },
  directives,
  sdls,
  services,
  onException: () => {
    // Disconnect from your database with an unhandled exception.
    db.$disconnect()
  },
  trustedDocuments: { store },
})
"
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql fragments are already setup updates the toml file with graphql and trusted documents enabled and keeps fragments 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[graphql]
  fragments = true
  trustedDocuments = true
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]
",
  "trustedDocumentsExists": false,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql fragments are already setup using no spaces updates the toml file with graphql and trusted documents enabled and keeps fragments 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[graphql]
  fragments=true
  trustedDocuments = true
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]
",
  "trustedDocumentsExists": false,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents and fragments are already setup makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents are already setup makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents are already setup using no spaces makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where no graphql or trusted documents is setup updates the toml file with graphql and trusted documents enabled 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]

[graphql]
  trustedDocuments = true",
  "trustedDocumentsExists": false,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates toml where graphql section is commented out adds a new section with \`trustedDocuments = true\` 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
# [graphql]
#  trustedDocuments = true
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]

[graphql]
  trustedDocuments = true",
  "trustedDocumentsExists": false,
}
`;
